File: packages/xyd-openapi2opencli/src/response.ts
import type { OpenAPIV3 } from 'openapi-types';
import type { XOpenApiResponse } from '@xyd-js/opencli';

import { arrayItems, getDefault, mergeAllOf, type Schema } from './schema';
import type { OpenApi2OpenCliOptions } from './types';

// Cap recursion so deep/recursive (dereferenced, possibly circular) response
// schemas can't produce an unbounded example.
const MAX_SAMPLE_DEPTH = 8;

/** True for a JSON-ish media type the schema sampler can faithfully represent. */
function isJsonMediaType(mediaType: string): boolean {
  return mediaType === 'application/json' || mediaType.endsWith('+json');
}

/** Pick the response media type to render (prefer JSON). */
function pickResponseContent(
  content: { [media: string]: OpenAPIV3.MediaTypeObject } | undefined,
): { mediaType: string; media: OpenAPIV3.MediaTypeObject } | undefined {
  if (!content) return undefined;
  const keys = Object.keys(content);
  if (!keys.length) return undefined;
  const json = keys.find((k) => k === 'application/json') || keys.find((k) => k.endsWith('+json'));
  const mediaType = json || keys[0];
  return { mediaType, media: content[mediaType] };
}

/**
 * A representative example value for a (dereferenced) response schema — curated
 * `example`/`default`/`enum` win, then per-type placeholders. Kept self-contained
 * (no external sampler dep) and bounded by depth.
 */
function sampleFromSchema(schema: Schema | undefined, depth = 0): unknown {
  if (!schema || depth > MAX_SAMPLE_DEPTH) return undefined;

  if ((schema as { example?: unknown }).example !== undefined) return (schema as { example?: unknown }).example;
  const def = getDefault(schema);
  if (def !== undefined) return def;
  if (Array.isArray(schema.enum) && schema.enum.length) return schema.enum[0];

  const merged = mergeAllOf(schema) ?? schema;

  const branches = (merged.oneOf || merged.anyOf) as Schema[] | undefined;
  if (Array.isArray(branches) && branches.length) return sampleFromSchema(branches[0], depth + 1);

  if (merged.type === 'array') {
    const item = sampleFromSchema(arrayItems(merged), depth + 1);
    return item === undefined ? [] : [item];
  }

  if (merged.type === 'object' || merged.properties) {
    const obj: Record<string, unknown> = {};
    for (const [key, raw] of Object.entries(merged.properties || {})) {
      if (key.startsWith('__')) continue; // dereferencer markers (__UNSAFE_*)
      const value = sampleFromSchema(raw as Schema, depth + 1);
      if (value !== undefined) obj[key] = value;
    }
    return obj;
  }

  switch (merged.type) {
    case 'string':
      return merged.format ? `<${merged.format}>` : 'string';
    case 'integer':
    case 'number':
      return 0;
    case 'boolean':
      return true;
    default:
      return undefined;
  }
}

/**
 * The concrete example body for a response media object. A curated `example` /
 * named example wins; otherwise the schema is sampled only for JSON-ish media
 * types (the sampler produces JSON-shaped values, so sampling an XML/binary body
 * would be misleading).
 */
function exampleFor(media: OpenAPIV3.MediaTypeObject, jsonish: boolean): unknown {
  if (media.example !== undefined) return media.example;
  const named = media.examples;
  if (named && typeof named === 'object') {
    const first = Object.values(named)[0] as OpenAPIV3.ExampleObject | undefined;
    if (first && 'value' in first && first.value !== undefined) return first.value;
  }
  return jsonish ? sampleFromSchema(media.schema as Schema | undefined) : undefined;
}

/**
 * Map an operation's success responses to the `x-openapi.responses` binding so
 * api.cli pages can render a response sample (mirrors the OpenAPI response
 * example track). Captures the 2xx responses that carry content; falls back to
 * the `default` response, then the first response with content. Only the example
 * is emitted (not the full schema) to keep generated OpenCLI docs lean.
 */
export function mapResponses(
  responses: OpenAPIV3.ResponsesObject | undefined,
  _options: OpenApi2OpenCliOptions,
): XOpenApiResponse[] {
  if (!responses) return [];

  const withContent = Object.entries(responses).filter(
    ([, r]) => r && typeof r === 'object' && (r as OpenAPIV3.ResponseObject).content,
  ) as [string, OpenAPIV3.ResponseObject][];
  if (!withContent.length) return [];

  // 2xx (incl. the OpenAPI "2XX" range wildcard) is success; else the documented
  // `default`; else the first content-bearing response.
  const success = withContent.filter(([status]) => /^2(\d\d|XX)$/i.test(status)).sort(([a], [b]) => a.localeCompare(b));
  const fallback = withContent.filter(([status]) => status === 'default');
  const chosen = success.length ? success : fallback.length ? fallback : [withContent[0]];

  const out: XOpenApiResponse[] = [];
  for (const [status, response] of chosen) {
    const picked = pickResponseContent(response.content);
    if (!picked) continue;
    const example = exampleFor(picked.media, isJsonMediaType(picked.mediaType));
    if (example === undefined || example === null) continue;

    const res: XOpenApiResponse = { status, contentType: picked.mediaType };
    if (response.description) res.description = response.description;
    res.example = example;
    out.push(res);
  }
  return out;
}

File: crates/xyd_openapi2opencli/src/response.rs
//! mapResponses + sampleFromSchema — port of response.ts. Emits the
//! x-openapi.responses example binding for the chosen success response.
//! Schemas already `ctx.resolve()`d (the caller resolves before sampling).

use serde_json::{Map, Value};
use std::collections::HashSet;

use crate::jsrt::js_object_keys;
use crate::model::XOpenApiResponse;
use crate::schema::{array_items, get_default, merge_all_of};

const MAX_SAMPLE_DEPTH: usize = 8;

fn is_json_media_type(m: &str) -> bool {
    m == "application/json" || m.ends_with("+json")
}

/// A representative example value for a schema (curated example/default/enum
/// win, then per-type placeholders). Bounded by depth.
fn sample_from_schema(schema: Option<&Value>, depth: usize) -> Option<Value> {
    let schema = schema?;
    if depth > MAX_SAMPLE_DEPTH {
        return None;
    }
    if let Some(ex) = schema.get("example") {
        return Some(ex.clone());
    }
    if let Some(def) = get_default(Some(schema)) {
        return Some(def.clone());
    }
    if let Some(Value::Array(e)) = schema.get("enum") {
        if !e.is_empty() {
            return Some(e[0].clone());
        }
    }

    let mut seen = HashSet::new();
    let merged = merge_all_of(schema, &mut seen);

    let branches = merged
        .get("oneOf")
        .or_else(|| merged.get("anyOf"))
        .and_then(|b| b.as_array());
    if let Some(branches) = branches {
        if !branches.is_empty() {
            return sample_from_schema(Some(&branches[0]), depth + 1);
        }
    }

    let type_str = merged.get("type").and_then(|t| t.as_str());
    if type_str == Some("array") {
        let item = sample_from_schema(array_items(Some(&merged)), depth + 1);
        return Some(match item {
            Some(v) => Value::Array(vec![v]),
            None => Value::Array(vec![]),
        });
    }

    if type_str == Some("object") || merged.get("properties").is_some() {
        let mut obj = Map::new();
        if let Some(props) = merged.get("properties").and_then(|p| p.as_object()) {
            for key in js_object_keys(props) {
                if key.starts_with("__") {
                    continue;
                }
                if let Some(v) = sample_from_schema(props.get(key), depth + 1) {
                    obj.insert(key.clone(), v);
                }
            }
        }
        return Some(Value::Object(obj));
    }

    match type_str {
        Some("string") => Some(Value::String(
            merged
                .get("format")
                .and_then(|f| f.as_str())
                .map(|f| format!("<{f}>"))
                .unwrap_or_else(|| "string".to_string()),
        )),
        Some("integer") | Some("number") => Some(Value::Number(0.into())),
        Some("boolean") => Some(Value::Bool(true)),
        _ => None,
    }
}

fn pick_response_content(content: Option<&Value>) -> Option<(String, Value)> {
    let map = content?.as_object()?;
    let keys = js_object_keys(map);
    if keys.is_empty() {
        return None;
    }
    let json = keys
        .iter()
        .find(|k| k.as_str() == "application/json")
        .or_else(|| keys.iter().find(|k| k.ends_with("+json")));
    let media_type = json
        .map(|k| (*k).clone())
        .unwrap_or_else(|| keys[0].clone());
    Some((media_type.clone(), map[&media_type].clone()))
}

fn example_for(media: &Value, jsonish: bool) -> Option<Value> {
    if let Some(ex) = media.get("example") {
        return Some(ex.clone());
    }
    if let Some(named) = media.get("examples").and_then(|e| e.as_object()) {
        // Object.values(named)[0]
        if let Some(first_key) = js_object_keys(named).into_iter().next() {
            if let Some(v) = named[first_key].get("value") {
                return Some(v.clone());
            }
        }
    }
    if jsonish {
        sample_from_schema(media.get("schema"), 0)
    } else {
        None
    }
}

fn is_2xx(status: &str) -> bool {
    // /^2(\d\d|XX)$/i
    let b = status.as_bytes();
    if b.len() != 3 || b[0] != b'2' {
        return false;
    }
    (b[1].is_ascii_digit() && b[2].is_ascii_digit())
        || ((b[1] == b'X' || b[1] == b'x') && (b[2] == b'X' || b[2] == b'x'))
}

/// `responses` is the (resolved) ResponsesObject.
pub fn map_responses(responses: Option<&Value>) -> Vec<XOpenApiResponse> {
    let Some(responses) = responses.and_then(|r| r.as_object()) else {
        return vec![];
    };

    // withContent, in Object.entries order.
    let with_content: Vec<(&String, &Value)> = js_object_keys(responses)
        .into_iter()
        .filter_map(|k| {
            let r = &responses[k];
            if r.is_object() && r.get("content").is_some() {
                Some((k, r))
            } else {
                None
            }
        })
        .collect();
    if with_content.is_empty() {
        return vec![];
    }

    let mut success: Vec<(&String, &Value)> = with_content
        .iter()
        .copied()
        .filter(|(s, _)| is_2xx(s))
        .collect();
    success.sort_by_key(|(a, _)| *a);
    let fallback: Vec<(&String, &Value)> = with_content
        .iter()
        .copied()
        .filter(|(s, _)| s.as_str() == "default")
        .collect();

    let chosen: Vec<(&String, &Value)> = if !success.is_empty() {
        success
    } else if !fallback.is_empty() {
        fallback
    } else {
        vec![with_content[0]]
    };

    let mut out = Vec::new();
    for (status, response) in chosen {
        let Some((media_type, media)) = pick_response_content(response.get("content")) else {
            continue;
        };
        let example = example_for(&media, is_json_media_type(&media_type));
        let Some(example) = example else { continue };
        if example.is_null() {
            continue;
        }
        out.push(XOpenApiResponse {
            status: (*status).clone(),
            content_type: media_type,
            description: response
                .get("description")
                .and_then(|d| d.as_str())
                .filter(|d| !d.is_empty())
                .map(|d| d.to_string()),
            example,
        });
    }
    out
}
